File: proxyshield/src/algorithms/sliding-window.js
/**
 * Sliding Window Log rate limiting algorithm.
 * Stores the timestamp of every request per IP+endpoint.
 * Counts how many timestamps fall within the current window.
 * Stricter than token bucket — no bursts allowed.
 */
export class SlidingWindow {
  /** Create a new SlidingWindow instance. */
  constructor() {
    /** @type {Map<string, number[]>} */
    this._windows = new Map();
  }

  /**
   * Check if a request is allowed under the sliding window algorithm.
   * @param {string} key - Rate limit key (e.g. "192.168.1.5:POST:/login")
   * @param {number} limit - Maximum requests allowed in the window
   * @param {number} windowSeconds - Window duration in seconds
   * @returns {{ allowed: boolean, current: number, limit: number, remaining: number, resetTime: number }}
   */
  check(key, limit, windowSeconds) {
    const now = Date.now();
    const windowStart = now - (windowSeconds * 1000);

    let entry = this._windows.get(key) || [];

    // Filter out expired timestamps
    entry = entry.filter(ts => ts > windowStart);

    const current = entry.length;

    if (current < limit) {
      entry.push(now);
      this._windows.set(key, entry);
      return {
        allowed: true,
        current: current + 1,
        limit,
        remaining: limit - current - 1,
        resetTime: entry[0] + (windowSeconds * 1000)
      };
    }

    // Rate limited
    const resetTime = entry[0] + (windowSeconds * 1000);
    this._windows.set(key, entry);
    return {
      allowed: false,
      current,
      limit,
      remaining: 0,
      resetTime
    };
  }

  /**
   * Remove expired entries to prevent memory leaks.
   * Call this periodically.
   */
  cleanup() {
    const now = Date.now();
    for (const [key, entry] of this._windows) {
      const filtered = entry.filter(ts => ts > now - 120_000);
      if (filtered.length === 0) {
        this._windows.delete(key);
      } else {
        this._windows.set(key, filtered);
      }
    }
  }
}

File: proxyshield/proxyshield-core/internal/algorithm/slidingwindow.go
package algorithm

import (
	"sync"
	"time"
)

// SlidingWindowEntry holds the per-key state for the sliding window algorithm.
// Uses a fixed-size bucket array — O(W) memory regardless of traffic volume.
type SlidingWindowEntry struct {
	mu         sync.Mutex
	buckets    []int32
	windowSize int
	lastSecond int64
	lastAccess time.Time
}

// SlidingWindow implements the counter-based sliding window rate limiting algorithm.
// Each second maps to one bucket in a fixed circular array — no timestamp logging.
// Memory usage is constant: 60 seconds = 60 int32s = 240 bytes per key.
type SlidingWindow struct {
	entries sync.Map // map[string]*SlidingWindowEntry
}

// NewSlidingWindow creates a new SlidingWindow instance.
func NewSlidingWindow() *SlidingWindow {
	return &SlidingWindow{}
}

// Check determines whether the key is within its rate limit.
// limit is the maximum number of requests per windowSeconds.
func (sw *SlidingWindow) Check(key string, limit int, windowSeconds int) RateLimitResult {
	now := time.Now()
	nowSec := now.Unix()

	actual, loaded := sw.entries.LoadOrStore(key, &SlidingWindowEntry{
		buckets:    make([]int32, windowSeconds),
		windowSize: windowSeconds,
		lastSecond: nowSec,
		lastAccess: now,
	})
	entry := actual.(*SlidingWindowEntry)

	if !loaded {
		// freshly created entry
		entry.mu.Lock()
		defer entry.mu.Unlock()
		idx := int(nowSec % int64(windowSeconds))
		entry.buckets[idx]++
		entry.lastAccess = now
		return RateLimitResult{
			Allowed:   1 <= limit,
			Current:   1,
			Limit:     limit,
			Remaining: limit - 1,
			ResetTime: nowSec + int64(windowSeconds),
		}
	}

	entry.mu.Lock()
	defer entry.mu.Unlock()

	entry.lastAccess = now

	// Clear stale buckets from lastSecond+1 up to now.
	if nowSec > entry.lastSecond {
		diff := nowSec - entry.lastSecond
		if diff > int64(windowSeconds) {
			diff = int64(windowSeconds)
		}
		for i := int64(1); i <= diff; i++ {
			staleSec := entry.lastSecond + i
			idx := int(staleSec % int64(windowSeconds))
			entry.buckets[idx] = 0
		}
		entry.lastSecond = nowSec
	}

	// Sum all buckets for current count.
	var total int32
	for _, v := range entry.buckets {
		total += v
	}
	current := int(total)

	resetTime := nowSec + int64(windowSeconds)

	if current < limit {
		idx := int(nowSec % int64(windowSeconds))
		entry.buckets[idx]++
		return RateLimitResult{
			Allowed:   true,
			Current:   current + 1,
			Limit:     limit,
			Remaining: limit - current - 1,
			ResetTime: resetTime,
		}
	}

	return RateLimitResult{
		Allowed:   false,
		Current:   current,
		Limit:     limit,
		Remaining: 0,
		ResetTime: resetTime,
	}
}

// Cleanup removes entries where lastSecond is older than maxAge.
// Call this periodically to prevent unbounded memory growth.
func (sw *SlidingWindow) Cleanup(maxAge time.Duration) {
	cutoff := time.Now().Add(-maxAge)
	sw.entries.Range(func(k, v interface{}) bool {
		entry := v.(*SlidingWindowEntry)
		entry.mu.Lock()
		old := entry.lastAccess.Before(cutoff)
		entry.mu.Unlock()
		if old {
			sw.entries.Delete(k)
		}
		return true
	})
}
